### PR TITLE
Starts implementation of Tuples class.

### PR DESCRIPTION
**Pull request by bsonntag**
`Tuples` is a utility class with factory methods as well as methods for the following operations:

- `flatRight` and `flatLeft`;
- `associateRight` and `associateLeft`;
- `map`;
- `zip`.

### CODE REVIEW

File: jtuples/src/test/java/org/jtuples/TuplesTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                new Pair<>("3", 3),
+                new Pair<>("4", 4)
+        );
+        
+        Iterator<Pair<String, Integer>> iterator =
+                Tuples.zip(firsts.iterator(), seconds.iterator());
+        
+        expected.forEach(pair -> assertEquals(pair, iterator.next()));
+    }
+
```
Comment (git-afsantos) [inline jtuples/src/test/java/org/jtuples/TuplesTest.java]: _It is obviously not the case, but anyway, just for theoretical correctness..._
What if `iterator` has more elements than `expected`? The assertions do not cover that case.

File: jtuples/src/main/java/org/jtuples/Tuples.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    
+    
+    public static <A, B, C>
+            Triple<A, B, C> flatFirst(Pair<Pair<A, B>, C> pair) {
+        return new Triple<>(
+                pair.first().first(),
+                pair.first().second(),
+                pair.second()
+        );
+    }
```
Comment (git-afsantos) [inline jtuples/src/main/java/org/jtuples/Tuples.java]: /**
     * Flattens a nested tuple, resulting in a tuple containing all the previous elements,
     * preserving order, but without the inner tuple on the first position.
     * That is, for a pair {@code ((a, b), c)} the result is the triple {@code (a, b, c)}.
     */

File: jtuples/src/main/java/org/jtuples/Tuples.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    
+    
+    public static <A, B, C>
+            Pair<Pair<A, B>, C> associateFirst(Pair<A, Pair<B, C>> pair) {
+        return new Pair<>(
+                new Pair<>(pair.first(), pair.second().first()),
+                pair.second().second()
+        );
+    }
```
Comment (git-afsantos) [inline jtuples/src/main/java/org/jtuples/Tuples.java]: /**
     * Associates a nested tuple on the first position, resulting in a tuple containing all the previous elements,
     * preserving order, but moving the position of the nesting.
     * That is, for a pair {@code (a, (b, c))} the result is the pair {@code ((a, b), c)}.
     */

File: jtuples/src/main/java/org/jtuples/Tuples.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    
+    
+    public static <A, B> Pair<B, B> map(Pair<? extends A, ? extends A> pair,
+            Function<A, B> f) {
+        return new Pair<>(
+                f.apply(pair.first()),
+                f.apply(pair.second())
+        );
+    }
```
Comment (git-afsantos) [inline jtuples/src/main/java/org/jtuples/Tuples.java]: /**
     * Maps a function over the given tuple, resulting in a new tuple containing the results
     * of applying the function to each tuple element, in order.
     * That is, for a function {@code f} and a pair {@code (a, b)} the result is the pair {@code (f(a), f(b))}.
     */

Comment (git-afsantos) [inline jtuples/src/main/java/org/jtuples/Tuples.java]: Or we could look into `List.replaceAll` for tips on documentation.

File: jtuples/src/main/java/org/jtuples/Tuples.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    
+    
+    public static <A, B> Collection<Pair<A, B>> zip(Iterable<A> firsts,
+            Iterable<B> seconds) {
+        Collection<Pair<A, B>> collection = new ArrayList<>();
+        zip(firsts.iterator(), seconds.iterator())
+                .forEachRemaining(collection::add);
+        return collection;
+    }
```
Comment (git-afsantos) [inline jtuples/src/main/java/org/jtuples/Tuples.java]: /**
     * Zips the given iterables, resulting in a collection containing pairs of all the previous elements,
     * preserving order, and stopping at the size of the smallest of the iterators.
     * That is, for lists {@code [a, b, c]} and {@code [d, e, f, g]} the result is the list of pairs
     * {@code [(a, d), (b, e), (c, f)]}.
     */

**Comment by git-afsantos** [comment]
The `Tuples` class can now include a `with` factory to produce a `Sextuple`, since they're implemented too.

**Comment by bsonntag** [comment]
> The `Tuples` class can now include a `with` factory to produce a `Sextuple`, since they're implemented too.

Done, and added a `map`, too.

**Comment by bsonntag** [comment]
@git-afsantos How is this looking? There's still some `@param`s, `@return`s and `@throws` missing.

File: jtuples/src/test/java/org/jtuples/TuplesTest.java (hunk 1 above)
Comment (bsonntag) [inline jtuples/src/test/java/org/jtuples/TuplesTest.java]: Good point. What do you suggest then? How could we check if two iterators are equal?

Comment (git-afsantos) [inline jtuples/src/test/java/org/jtuples/TuplesTest.java]: @bsonntag We should extract an iterator from the list, and then we must guarantee that their elements and their sizes are equal.

    Iterator<Pair<String, Integer>> expectedIterator = expected.iterator();

    while (expectedIterator.hasNext()) {
        assertEquals(expectedIterator.next(), iterator.next());
    }
    assertEquals(expectedIterator.hasNext(), iterator.hasNext());

This covers empty iterators, equal sizes, different sizes, and different elements. It's all we need.

**Comment by git-afsantos** [comment]
@bsonntag So far, it's looking good to me.

File: jtuples/src/test/java/org/jtuples/TuplesTest.java (hunk 1 above)
Comment (bsonntag) [inline jtuples/src/test/java/org/jtuples/TuplesTest.java]: I like it. Done.

**Comment by git-afsantos** [comment]
I just noticed that, as it stands, `Tuples` can be instantiated. These static and factory classes usually go along with a private constructor.

    private Tuples() {
        throw new AssertionError("This class should not be instantiated.");
    }

**Comment by bsonntag** [comment]
It already has a private constructor, it just doesn't throw an error. I'll change it to do that.

**Comment by git-afsantos** [comment]
Somehow I missed the private constructor, hence my comment.

**Comment by git-afsantos** [comment]
Looks good to me. :+1:

**Pull request merged**